### ISSUE TITLE
support src being an empty string

### ISSUE THREAD

**Issue by gkatsev** [videojs/video.js#2852]
the spec basically says that when the src is set to an empty string, you want to reset the player and then stop and not try to loading anything. See step 9 of http://www.w3.org/TR/html5/embedded-content-0.html#concept-media-load-algorithm

**Comment by Soviut** [comment]
This is currently how I test my error styles. Perhaps it would be a good idea to document how to put the player into an error state for error testing.

**Comment by gkatsev** [comment]
You can trigger an error by calling `player.error()` with an object that has a `code` property and various others: `player.error({code: 4})`.

**Comment by Soviut** [comment]
Ah nice, I assumed there was some way to do it; I should have guessed it was that simple.

**Comment by gkatsev** [comment]
So, after trying playing around with setting the `src` to `''`, I realized that it does throw an error. Tested on this page: http://www.w3.org/2010/05/video/mediaevents.html
However, setting the `src` to `null` didn't throw and that might be an acceptable use.
Also, perhaps we want something like a `player.unload()` instead of setting the `src` to `null`?

**Comment by Soviut** [comment]
Could we not just assume that an empty call to `player.src()` means remove call sources?

Also, I'd expect `player.unload()` to remove videoJS from the target and return the `<video>` tag to its original state, not change sources.

**Comment by gkatsev** [comment]
@Soviut `player.src()` needs to act as a getter.
I guess `unload` isn't the right word for it. Just not sure that `player.src(null)` is the best way of implementing the user facing API for this.

**Comment by Soviut** [comment]
Perhaps just document that `player.src({})` or `player.src([])` could clear the sources.

**Comment by gkatsev** [comment]
Well, what I specifically want to happen with `src(null)` is to change the tech to Html5, and put in there a simple video element with no source. Basically, unset all state of the player as much as possible.

**Comment by Soviut** [comment]
That sounds more like a reset. Perhaps a `player.reset()` or `player.init()` is a good idea and then that can be called if src is set to null as well.

**Comment by dmlap** [comment]
If facilities exist in the HTML spec to "reset the video element" than I think we should stick to those alone. Having two ways of doing the same thing can be confusing.

**Comment by gkatsev** [comment]
Trying to read the spec to see if there's a "reset" functionality. Seems like it might be just removing all source elements, deleting the src attribute, and then calling `load()`. Which is what I have hacked together locally currently.

**Comment by misteroneill** [comment]
I think `player.reset()` makes sense in that case. It's more meaningful to me than `player.src(null)`.

**Comment by gkatsev** [comment]
Actually, I found the spec say exactly that: http://www.w3.org/TR/html5/embedded-content-0.html#best-practices-for-authors-using-media-elements
> emoving the element's src attribute and any source element descendants, and invoking the element's load() method.

**Comment by dmlap** [comment]
Well, questions about reset() aside, sounds like we should support nulling the src and calling load()

**Comment by gkatsev** [comment]
As in:
```js
player.src(null);
player.load();
```?
Except it would cause an error to load because it will translate to
``js
video.src = null;
video.load();
```
Which does cause an error. Unless we want special handling for `src(null)` which would clear out the source element and delete the attribute?

**Comment by dmlap** [comment]
Ah, I overlooked the removal aspect. Tricky.

**Comment by gkatsev** [comment]
We've decided to go with `Player#reset` unless we figure out a better name for this before we merge in the PR :)

**Comment by gkatsev** [comment]
I have made a PR that adds the `reset` method to the player. Check it out in #2880.

**Comment by i18n-hoon** [comment]
I need help with this issue.
I want to use videojs as a local video player.
If I set the valid video source and change the local video file, there is no problem.

However, if I set the source to null initially and change the video file, it will not play.
The video is loaded successfully and seeking is possible, but It can't play. And also not work in programming api (`player.play ()`)

This is my code how i change the video file.

```javascript
const file = e.target.files[0];
const source = URL.createObjectURL(new Blob([file]));
player.src([{type: 'video/mp4', source}]);
```
